src/LabelGroup/LabelGroup.tsx

```diff
@@ -1,17 +1,365 @@
+  children,
+  visibleChildCount,
+  overflowStyle,
+  sx: sxProp,
+}) => {
+  const containerRef = React.useRef<HTMLDivElement>(null)
+  const collapseButtonRef = React.useRef<HTMLButtonElement>(null)
+  const firstHiddenIndexRef = React.useRef<number | undefined>(undefined)
+  const [visibilityMap, setVisibilityMap] = React.useState<Record<string, boolean>>({})
+  const [isOverflowShown, setIsOverflowShown] = React.useState<boolean>(false)
+  const [buttonClientRect, setButtonClientRect] = React.useState<DOMRect>({
+    width: 0,
+    right: 0,
+    height: 0,
+    x: 0,
+    y: 0,
+    top: 0,
+    left: 0,
+    bottom: 0,
+    toJSON: () => undefined,
+  })
+
+  const {theme} = useTheme()
+
+  const overlayPaddingPx = parseInt(get('space.2')(theme), 10)
+
+  const hiddenItemIds = Object.keys(visibilityMap).filter(key => !visibilityMap[key])
+
+  // `overlayWidth` is only needed when we render an overlay
+  // if we don't use an overlay, we can skip the width calculation
+  // and save on reflows caused by measuring DOM nodes.
+  const overlayWidth =
+    visibleChildCount === 'auto' && overflowStyle === 'overlay'
+      ? getOverlayWidth(buttonClientRect, containerRef, overlayPaddingPx)
+      : undefined
+
+  const expandButtonRef: React.RefCallback<HTMLButtonElement> = React.useCallback(
+    node => {
+      if (node !== null) {
+        const nodeClientRect = node.getBoundingClientRect()
+
+        if (nodeClientRect.width !== buttonClientRect.width || nodeClientRect.right !== buttonClientRect.right) {
+          setButtonClientRect(nodeClientRect)
+        }
+
+        // @ts-ignore you can set `.current` on ref objects or ref callbacks in React
+        expandButtonRef.current = node
+      }
+    },
+    [buttonClientRect],
+  )
+
+  // Sets the visibility map to hide children after the given index.
+  const hideChildrenAfterIndex = React.useCallback((truncateAfter: number) => {
+    const containerChildren = containerRef.current?.children || []
+    const updatedEntries: Record<string, boolean> = {}
+    for (const child of containerChildren) {
+      const targetId = child.getAttribute('data-index')
+      if (targetId) {
+        updatedEntries[targetId] = parseInt(targetId, 10) < truncateAfter
+      }
+    }
+
+    setVisibilityMap(updatedEntries)
+  }, [])
+
+  const openOverflowOverlay = React.useCallback(() => setIsOverflowShown(true), [setIsOverflowShown])
+
+  const closeOverflowOverlay = React.useCallback(() => {
+    setIsOverflowShown(false)
+  }, [setIsOverflowShown])
+
+  const collapseInlineExpandedChildren = React.useCallback(() => {
+    setIsOverflowShown(false)
+
+    if (visibleChildCount && typeof visibleChildCount === 'number') {
+      hideChildrenAfterIndex(visibleChildCount)
+    }
+
+    // We need to manually re-focus the collapse button if we're not showing the full
+    // list in an overlay.
+    // TODO: get rid of this hack
+    setTimeout(() => {
```

Comment:
Help wanted: I used this `setTimeout` because we need to wait until state has been updated and the expand button is back in the DOM.

This feels sloppy/hacky, but I haven't been able to figure out a better way. @siddharthkp - we paired on this last week.